middleware/complexconfig/provider.go

```diff
@@ -0,0 +1,181 @@
+		}
+		var spec schema.PackageSpec
+		if err := json.Unmarshal([]byte(schemaResp.Schema), &spec); err != nil {
+			return p.CheckResponse{},
+				p.InternalErrorf("unable to decode config: invalid schema: %w", err)
+		}
+
+		for k, spec := range spec.Config.Variables {
+			v, ok := req.News[resource.PropertyKey(k)]
+			if !ok {
+				continue
+			}
+
+			req.News[resource.PropertyKey(k)] = fixEncoding(v, spec)
+		}
+
+		return check(ctx, req)
+	}
+}
+
+func fixEncoding(v resource.PropertyValue, spec schema.PropertySpec) (out resource.PropertyValue) {
+	// Ensure that v is unwrapped
+
+	if putil.IsComputed(v) {
+		return v
+	}
+
+	if putil.IsSecret(v) {
+		defer func() {
+			out = putil.MakeSecret(v)
+		}()
+		v = putil.MakePublic(v)
+	}
+
+	// If the value is not a string, we assume that it is the correct type. If the
+	// value is string-typed in the spec, there's no point unmarshaling it since it's
+	// a primitive string.
+	if !v.IsString() || spec.Type == "string" {
+		return v
+	}
+
+	var target any
+	err := json.Unmarshal([]byte(v.StringValue()), &target)
+	if err != nil {
+		return v
+	}
+
+	// Instead of using resource.NewPropertyValue, specialize it to detect nested json-encoded secrets.
+	var replv func(encoded any) (resource.PropertyValue, bool)
+	replv = func(v any) (resource.PropertyValue, bool) {
+		if s, ok := v.(string); ok {
+			switch s {
+			case plugin.UnknownBoolValue:
+				return resource.MakeComputed(resource.NewProperty(false)), true
+			case plugin.UnknownNumberValue:
+				return resource.MakeComputed(resource.NewProperty(0.0)), true
+			case plugin.UnknownStringValue:
+				return resource.MakeComputed(resource.NewProperty("")), true
+			case plugin.UnknownArrayValue:
+				return resource.MakeComputed(resource.NewProperty([]resource.PropertyValue{})), true
+			case plugin.UnknownObjectValue:
+				return resource.MakeComputed(resource.NewProperty(resource.PropertyMap{})), true
+			case plugin.UnknownAssetValue:
+				return resource.MakeComputed(resource.NewProperty(&resource.Asset{})), true
+			case plugin.UnknownArchiveValue:
+				return resource.MakeComputed(resource.NewProperty(&resource.Archive{})), true
+			}
+		}
+		m, ok := v.(map[string]any)
+		if !ok {
+			return resource.PropertyValue{}, false
+		}
+
+		value, ok := m[sig.Key]
+		if !ok {
+			return resource.PropertyValue{}, false
+		}
+		sigValue, ok := value.(string)
+		if !ok {
+			return resource.PropertyValue{}, false
+		}
+
+		switch sigValue {
+		case sig.Secret:
+			return putil.MakeSecret(
+				resource.NewPropertyValueRepl(m["value"], nil, replv),
+			), true
+		case sig.OutputValue:
+			castBool := func(key string) bool {
+				v, ok := m[key]
+				if !ok {
+					return false
+				}
+				b, ok := v.(bool)
+				return ok && b
+			}
+
+			elem, hasElem := m["value"]
+
+			return resource.NewProperty(resource.Output{
+				Known:        hasElem,
+				Secret:       castBool("secret"),
+				Dependencies: nil,
```

Comment:
Always nil?

Reply:
No. Good call. The rapid test property values didn't generate interesting dependency values. I addressed this and added valid support for dependency values in a subsequent push.